llrb Floor function can occur nil pointer dereference

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!
-->


**What happened**:
Current LLRB Floor() implementation can make error in case of below simple scenario.
```go
tree := llrb.NewTree[*intKey, *intValue]()
tree.Put(newIntKey(10), newIntValue(10))
_, v := tree.Floor(newIntKey(5))
assert.Nil(t, v)
```
It has only root node, so parent is nil -> returning parent.key and parent.value makes nil pointer dereference.

```
--- FAIL: TestTree (0.00s)
    --- FAIL: TestTree/floor_test (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered, repanicked]
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x100720a5c]

goroutine 6 [running]:
testing.tRunner.func1.2({0x1007c8ee0, 0x10098ced0})
        /opt/homebrew/Cellar/go/1.25.1/libexec/src/testing/testing.go:1872 +0x190
testing.tRunner.func1()
        /opt/homebrew/Cellar/go/1.25.1/libexec/src/testing/testing.go:1875 +0x31c
panic({0x1007c8ee0?, 0x10098ced0?})
        /opt/homebrew/Cellar/go/1.25.1/libexec/src/runtime/panic.go:783 +0x120
github.com/yorkie-team/yorkie/pkg/llrb.(*Tree[...]).Floor(0x10080f900?, 0x1400000f708)
        /Users/ggyuchive/yorkieProject/yorkie/pkg/llrb/llrb.go:127 +0x11c
github.com/yorkie-team/yorkie/pkg/llrb_test.TestTree.func2(0x14000003dc0)
        /Users/ggyuchive/yorkieProject/yorkie/pkg/llrb/llrb_test.go:114 +0x298
testing.tRunner(0x14000003dc0, 0x1400007c940)
        /opt/homebrew/Cellar/go/1.25.1/libexec/src/testing/testing.go:1934 +0xc8
created by testing.(*T).Run in goroutine 4
        /opt/homebrew/Cellar/go/1.25.1/libexec/src/testing/testing.go:1997 +0x364
FAIL    github.com/yorkie-team/yorkie/pkg/llrb  0.576s
FAIL
```
**What you expected to happen**:
Successfully return nil, nil value on above test.

**How to reproduce it (as minimally and precisely as possible)**:
In `pkg/llrb/llrb_test.go`, add
```go
t.Run("floor test", func(t *testing.T) {
	tree := llrb.NewTree[*intKey, *intValue]()
	tree.Put(newIntKey(10), newIntValue(10))
	_, v := tree.Floor(newIntKey(5))
	assert.Nil(t, v)
})
```
and run `go test ./pkg/llrb`

**Anything else we need to know?**:
We can refine Floor function without using parent pointer.

**Environment**:
- Operating system:
- Browser and version:
- Yorkie version (use `yorkie version`):
- Yorkie JS SDK version:

--- a/pkg/llrb/llrb.go
+++ b/pkg/llrb/llrb.go
@@ -37,12 +37,11 @@ type Value interface {
 
 // Node is a node of Tree.
 type Node[K Key, V Value] struct {
-	key    K
-	value  V
-	parent *Node[K, V]
-	left   *Node[K, V]
-	right  *Node[K, V]
-	isRed  bool
+	key   K
+	value V
+	left  *Node[K, V]
+	right *Node[K, V]
+	isRed bool
 }
 
 // NewNode creates a new instance of Node.
@@ -101,39 +100,22 @@ func (t *Tree[K, V]) Remove(key K) {
 // Floor returns the greatest key less than or equal to the given key.
 func (t *Tree[K, V]) Floor(key K) (K, V) {
 	node := t.root
+	var resultK K
+	var resultV V
 
 	for node != nil {
-		compare := key.Compare(node.key)
-		if compare > 0 {
-			if node.right != nil {
-				node.right.parent = node
-				node = node.right
-			} else {
-				return node.key, node.value
-			}
-		} else if compare < 0 {
-			if node.left != nil {
-				node.left.parent = node
-				node = node.left
-			} else {
-				parent := node.parent
-				child := node
-				for parent != nil && child == parent.left {
-					child = parent
-					parent = parent.parent
-				}
-
-				// TODO(hackerwins): check below warning
-				return parent.key, parent.value
-			}
-		} else {
+		switch key.Compare(node.key) {
+		case 0:
 			return node.key, node.value
+		case -1:
+			node = node.left
+		case 1:
+			resultK, resultV = node.key, node.value
+			node = node.right
 		}
 	}
 
-	var zeroK K
-	var zeroV V
-	return zeroK, zeroV
+	return resultK, resultV
 }
 
 // Len returns the length of the tree.

--- a/pkg/llrb/llrb_test.go
+++ b/pkg/llrb/llrb_test.go
@@ -58,18 +58,38 @@ func (v *intValue) String() string {
 	return fmt.Sprintf("%d", v.value)
 }
 
-func TestTree(t *testing.T) {
-	t.Run("keeping order test", func(t *testing.T) {
-		arrays := [][]int{
-			{0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
-			{8, 5, 7, 9, 1, 3, 6, 0, 4, 2},
-			{7, 2, 0, 3, 1, 9, 8, 4, 6, 5},
-			{2, 0, 3, 5, 8, 6, 4, 1, 9, 7},
-			{8, 4, 7, 9, 2, 6, 0, 3, 1, 5},
-			{7, 1, 5, 2, 8, 6, 3, 4, 0, 9},
-			{9, 8, 7, 6, 5, 4, 3, 2, 1, 0},
+func checkFloor(t *testing.T, tree *llrb.Tree[*intKey, *intValue], arr []*intKey, loop int) {
+	for floorKey := range loop {
+		Key := newIntKey(floorKey)
+		var expectedKey *intKey
+		for _, v := range arr {
+			if v.key <= Key.key {
+				if expectedKey == nil || expectedKey.key <= v.key {
+					expectedKey = v
+				}
+			}
 		}
+		resultKey, _ := tree.Floor(Key)
 
+		if expectedKey == nil {
+			assert.Nil(t, resultKey)
+		} else {
+			assert.Equal(t, expectedKey.key, resultKey.key)
+		}
+	}
+}
+
+func TestTree(t *testing.T) {
+	arrays := [][]int{
+		{0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
+		{8, 5, 7, 9, 1, 3, 6, 0, 4, 2},
+		{7, 2, 0, 3, 1, 9, 8, 4, 6, 5},
+		{2, 0, 3, 5, 8, 6, 4, 1, 9, 7},
+		{8, 4, 7, 9, 2, 6, 0, 3, 1, 5},
+		{7, 1, 5, 2, 8, 6, 3, 4, 0, 9},
+		{9, 8, 7, 6, 5, 4, 3, 2, 1, 0},
+	}
+	t.Run("keeping order test", func(t *testing.T) {
 		for _, array := range arrays {
 			tree := llrb.NewTree[*intKey, *intValue]()
 			for _, value := range array {
@@ -85,6 +105,19 @@ func TestTree(t *testing.T) {
 
 			tree.Remove(newIntKey(5))
 			assert.Equal(t, "0,1,3,4,6,7,9", tree.String())
+		}
+	})
+
+	t.Run("floor test", func(t *testing.T) {
+		for _, array := range arrays {
+			testArr := []*intKey{}
+			tree := llrb.NewTree[*intKey, *intValue]()
+			for _, value := range array {
+				checkFloor(t, tree, testArr, 10)
+				tree.Put(newIntKey(value), newIntValue(value))
+				testArr = append(testArr, newIntKey(value))
+				checkFloor(t, tree, testArr, 10)
+			}
 		}
 	})
 }

